x-pack/libbeat/management/api/event_reporter.go

```diff
@@ -0,0 +1,131 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package api
+
+import (
+	"sync"
+	"time"
+
+	"github.com/joeshaw/multierror"
+
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+var debugK = "event_reporter"
+
+// EventReporter is an object that will periodically send asyncronously events to the
+// CM events endpoints.
+type EventReporter struct {
+	logger       *logp.Logger
+	client       AuthClienter
+	wg           sync.WaitGroup
+	events       chan Event
+	period       time.Duration
+	maxBatchSize int
+}
+
+// NewEventReporter returns a new event reporter
+func NewEventReporter(
+	logger *logp.Logger,
+	client AuthClienter,
+	period time.Duration,
+	maxBatchSize int,
+) *EventReporter {
+	log := logger.Named(debugK)
+	return &EventReporter{
+		logger:       log,
+		client:       client,
+		period:       period,
+		events:       make(chan Event),
+		maxBatchSize: maxBatchSize,
+	}
+}
+
+// Start starts the event reported and wait for new events.
+func (e *EventReporter) Start() {
+	e.wg.Add(1)
+	go e.worker()
+
+	e.logger.Info("Starting event reporter service")
+}
+
+// Stop stops the reporting events to the endpoint.
+func (e *EventReporter) Stop() {
+	e.logger.Info("Stopping event reporter service")
+	close(e.events)
+	e.wg.Wait()
+}
+
+func (e *EventReporter) worker() {
+	defer e.wg.Done()
+
+	ticker := time.NewTicker(e.period)
+	defer ticker.Stop()
+
+	var buffer []Event
+	for {
+		select {
+		case event, ok := <-e.events:
+			if !ok {
+				e.reportEvents(buffer)
+				return
+			}
+			buffer = append(buffer, event)
+		case <-ticker.C:
+			e.reportEvents(buffer)
+			buffer = nil
+		}
+	}
+}
+
+func (e *EventReporter) reportEvents(events []Event) {
+	if len(events) == 0 {
+		return
+	}
+	e.logger.Debug("Reporting %d events to Kibana", len(events))
+
+	// NOTE: Should we retry here? or do X attempts.
+	if err := e.sendBatchEvents(events); err != nil {
+		e.logger.Errorf("could not send events, error: %+v", err)
+	}
+}
+
+func (e *EventReporter) sendBatchEvents(events []Event) error {
+	var errors multierror.Errors
+	for pos := 0; pos < len(events); pos += e.maxBatchSize {
+		j := pos + e.maxBatchSize
+		if j > len(events) {
+			j = len(events)
+		}
+		if err := e.sendEvents(events[pos:j]); err != nil {
+			errors = append(errors, err)
+		}
+	}
```

Comment:
alternative implementations splitting the slice + using objects instead of indices:

```
for len(events) > e.maxBatchSize {
  batch := events[:e.maxBatchSize]
  events = events[e.maxBatchSize:]
  if err := e.sendEvents(batch); err != nil {
    errors = append(errors, err)
  }
}
if len(events) > 0 {
  if err := e.sendEvents(events); err != nil {
    errors = append(errors, err)
  }
}
```

or 

```
for len(events) > 0 {
  batch := events
  if len(batch) > e.maxBatchSize {
    batch = batch[:e.maxBatchSize]
  }

  if err := e.sendEvents(batch); err != nil {
    errors = append(errors, err)
  }

  events = events[len(batch):]
}
```

P.S: I don't mind using indices.